### PR TITLE
Fallback to local cache when there are issues connecting to the Redis cache

### PR DESCRIPTION
**Pull request by cds95**
## Description

This ticket implements the fallback method for adapters to start using their local in memory cache if the connection to Redis fails.

## Changes

- Implement fallback method to handle Redis cache failures
- Make local LRU cache singleton

## Steps to Test

Testing single requests

1. Run `docker compose up` from `/grafana` folder to start Redis cluster
2. Send requests to ensure that the adapter is caching requests correctly using Redis. 
3. Shutdown Redis
4. Send requests and ensure that requests were still being cached.

Testing WS requests

1. Enable websockets by running `export WS_ENABLED=true`
2. Follow the instructions to test single requests
3. Verify that the adapter saves WS results by searching for this log `WS: Saved result`.  Note the result.
4. Send another request to the adapter and verify that the value matches what is saved and that the response came from the cache.

Note:  You can tell that requests are being cached by turning setting `API_VERBOSE` to true and verifying that the `cacheHit` flag in the response has been set to true.

Example

```
{
    "jobRunID": "278c97ffadb54a5bbb93cfec5f7b5503",
    "result": 4310.24,
    "maxAge": 120000,
    "debug": {
        "cacheHit": true,
        "staleness": 0.789,
        "performance": 0.003814976,
        "providerCost": 0
    },
    "statusCode": 200,
    "data": {
        "result": 4310.24
    }
}
```

## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `<ADAPTER_PACKAGE>/schemas/env.json` and `<ADAPTER_PACKAGE>/README.md`
- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [x] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)

### CODE REVIEW

File: packages/core/bootstrap/src/lib/cache/index.ts [hunk 1]
```diff
@@ -231,36 +257,71 @@ export const withCache =
...
       })
 
       try {
-        const cachedAdapterResponse = await adapterCache.getResultForRequest(adapterRequest)
+        const cachedAdapterResponse = await adapterCacheToUse.getResultForRequest(adapterRequest)
         if (cachedAdapterResponse) return cachedAdapterResponse
       } catch (error) {
-        logger.warn(`Cache middleware error! Passing through. `, error)
-        return await execute(adapterRequest, context)
+        const localAdapterCache = new AdapterCache({
```
Comment (cds95) [inline packages/core/bootstrap/src/lib/cache/index.ts:285]: refactor duplicate code

File: packages/core/bootstrap/src/lib/cache/index.ts [hunk 2]
```diff
@@ -223,6 +228,27 @@ export class AdapterCache {
...
+  execute: Execute,
+  adapterCache: AdapterCache,
+): Promise<AdapterResponse | undefined> => {
+  const type = env.CACHE_TYPE || DEFAULT_CACHE_TYPE
+  if (type === 'local') {
+    logger.warn('Cache type already set to local.  Passing through...')
+    return await execute(adapterRequest, context)
+  }
+  try {
+    logger.warn(`Trying to fetch from local cache. `, error)
```
Comment (cds95) [inline packages/core/bootstrap/src/lib/cache/index.ts:244]: Remove trailing whitespace

**Comment by cds95** [comment]
Skipping writing tests for now as we should think about how to refactor the file containing the cache logic as well as setup mocks.

https://app.shortcut.com/chainlinklabs/story/23631/setup-mock-testing-for-cache

**Review by austinborn — APPROVED** [review]
LGTM

**Pull request merged**